jwst/pipeline/calwebb_tso3.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import unicode_literals, absolute_import
+
+import os
+
+from astropy.table import vstack
+
+from ..stpipe import Pipeline
+from .. import datamodels
+
+from ..outlier_detection import outlier_detection_tso_step
+from ..outlier_detection import outlier_detection_scaled_step
+from ..tso_photometry import tso_photometry_step
+from ..extract_1d import extract_1d_step
+from ..white_light import white_light_step
+
+__version__ = "0.7.8"
+
+
+class TSO3Pipeline(Pipeline):
+    """
+    TSO3Pipeline: Applies level 3 processing to TSO-mode data from
+                    any JWST instrument.
+
+    Included steps are:
+        outlier_detection
+        tso_photometry
+        extract_1d
+        white_light
+    """
+
+    spec = """
+        suffix = string(default='i2d')
+        scale_detection = boolean(default=False)
+    """
+
+    # Define alias to steps
+    step_defs = {'outlier_detection': 
+                        outlier_detection_tso_step.OutlierDetectionTSOStep,
+                 'outlier_detection_scaled': 
+                        outlier_detection_scaled_step.OutlierDetectionScaledStep,
+                 'tso_photometry': tso_photometry_step.TSOPhotometryStep,
+                 'extract_1d': extract_1d_step.Extract1dStep,
+                 'white_light': white_light_step.WhiteLightStep
+                 }
+
+    def process(self, input):
+        """
+        Run the TSO3Pipeline
+
+        Parameters
+        ----------
+        input: Level3 Association, or ModelContainer
+            The exposures to process
+        """
+
+        self.log.info('Starting calwebb_tso3 ...')
+        print("scale_detection set to : {}".format(self.scale_detection))
+        
+        input_asn = datamodels.open(input)
+        product_name = input_asn.meta.asn_table.products[0].name
+
+        # Setup output creation
+        #self.output_basename = product['name']
+        self.output_basename = product_name 
+        
+        # Input may consist of multiple exposures, so loop over each of them
+        for cube in input_asn:
+            # Convert CubeModel into ModelContainer of 2-D DataModels
+            input_models = datamodels.ModelContainer()
+            for i in range(cube.data.shape[0]):
+                # convert each plane of data cube into it own array
+                # for outlier detection...
+                image = datamodels.ImageModel(data=cube.data[i],
+                        err=cube.err[i], dq=cube.dq[i])
+                image.meta = cube.meta
+                input_models.append(image)
+            
+            if not self.scale_detection:
+                self.log.info("Performing outlier detection on input images...")
+                results = self.outlier_detection(input_models)
+            else:
+                self.log.info("Performing scaled outlier detection on input images...")
+                results = self.outlier_detection_scaled(input_models)
```

Comment:
How is `results` different than `input_models`? If `input_models` is unchanged, while `results` is a copy of the input with the DQ arrays updated, then line 90 below should be copying from `results[i].dq`, not `input_models[i].dq.` If `input_models` has updated DQ flags, then what is `results` used for?